include/deal.II/lac/trilinos_tpetra_vector.templates.h

```diff
@@ -411,6 +411,31 @@ namespace LinearAlgebra
 
 
 
+    template <typename Number, typename MemorySpace>
+    template <typename OtherNumber>
+    Vector<Number, MemorySpace> &
+    Vector<Number, MemorySpace>::operator=(const dealii::Vector<OtherNumber> &V)
+    {
+#  ifdef DEBUG
+      bool number_type_matches = std::is_same_v<OtherNumber, Number>;
+      Assert(number_type_matches, ExcInternalError());
+#  endif
+
+      vector.reset();
+      nonlocal_vector.reset();
+
+      Teuchos::Array<OtherNumber> vector_data(V.begin(), V.end());
+      vector = Utilities::Trilinos::internal::make_rcp<VectorType>(
+        V.locally_owned_elements().make_tpetra_map_rcp(), vector_data);
```

Comment:
Why don't we just implement this properly and do
```suggestion
      vector.reset();
      nonlocal_vector.reset();

       if constexpr(std::is_same_v<OtherNumber, Number>) {
        Teuchos::Array<OtherNumber> vector_data(V.begin(), V.end());
        vector = Utilities::Trilinos::internal::make_rcp<VectorType>(
          V.locally_owned_elements().make_tpetra_map_rcp(), vector_data);
       } else {
         vector = Utilities::Trilinos::internal::make_rcp<VectorType>(
          V.locally_owned_elements().make_tpetra_map_rcp(), /*zeroOut*/ false);
#  if DEAL_II_TRILINOS_VERSION_GTE(13, 2, 0)
          auto vector_2d = vector->template getLocalView<Kokkos::HostSpace>(
            Tpetra::Access::ReadWrite);
#  else
          vector->template sync<Kokkos::HostSpace>();
          auto vector_2d = vector->template getLocalView<Kokkos::HostSpace>();
#  endif
          auto vector_1d = Kokkos::subview(vector_2d, Kokkos::ALL(), 0);
#  if !DEAL_II_TRILINOS_VERSION_GTE(13, 2, 0)
          vector->template modify<Kokkos::HostSpace>();
#  endif
          const size_t localLength = vector->getLocalLength();
          for (size_t k = 0; k < localLength; ++k)
            vector_1d(k) = V(k);
#  if !DEAL_II_TRILINOS_VERSION_GTE(13, 2, 0)
          vector->template sync<
          typename Tpetra::Vector<Number, int, types::signed_global_dof_index>::
            device_type::memory_space>();
#  endif 
       } 
```
or do the copy on the device by using an unmanaged Kokkos::View for `dealii::Vector<OtherNumber> V`.

Reply:
This will probably not compile for mixed number types since the compiler will try to compile this function in any case (even though it is never called)
```
        vector = Utilities::Trilinos::internal::make_rcp<VectorType>(
          V.locally_owned_elements().make_tpetra_map_rcp(), vector_data);
```
And I suspect that this will lead to a linking error...